### ISSUE TITLE
Puma 6.0 - Out of sync bug or misconfiguration?

### ISSUE THREAD

**Issue by JensDebergh** [puma/puma#3022]
**Describe the bug**

I've upgraded from puma `5.6.4` to `6.0` on an application that is running for over 3 years in production and no logic regarding how the invoice is updated has changed in over a year.

At first everything seems fine when creating invoices via the editor, but at some point the server gets out of sync or is sharing memory with a previous request. Items that were not inputted are being added to the line items and the amount isn't calculated based on the items, which lets me to believe something is going on with Puma workers and threads.

It's like some memory of the previous request stays behind and gets added to the new request and vice versa. I have no special setup with puma or doing any funny business with configs.

Video:

Puma 6.0:

https://monosnap.com/file/f8Qezrlo3ynndFyVwi6aH1XH8QQ0ly

Puma 5.6.4

https://monosnap.com/file/WiltGS9B1oDFoksRIKqs41y1p1SWpY

This only happens on puma `6.0`. Reverting back to `5.6` resolves the issue but doesn't let me use the latest of puma webserver

**Puma config:**

Gemfile.lock

```
    puma (6.0.0)
      nio4r (~> 2.0)
```

```
workers Integer(ENV.fetch("WEB_CONCURRENCY", 3))
threads_count = Integer(ENV.fetch("MAX_THREADS", 2))
threads(threads_count, threads_count)

preload_app!

environment ENV.fetch("RACK_ENV", "development")

on_worker_boot do
  # Worker specific setup for Rails 4.1+
  # See: https://devcenter.heroku.com/articles/deploying-rails-applications-with-the-puma-web-server#on-worker-boot
  ActiveRecord::Base.establish_connection
end
```

**To Reproduce**

Hard to reproduce since I would need to share my project.

**Expected behavior**

I would expect it to run under the same conditions as `5.6.4`.

**Desktop (please complete the following information):**
 - OS: Mac & Linux (Issue arises on heroku too)
 - Puma Version [6.0]

**Comment by dentarg** [comment]
haven't looked at the videos but that doesn't sound good 😞

This may be https://github.com/puma/puma/issues/2999 that's has been addressed in the master branch. Are you able to test your app with the master branch of Puma?

**Comment by mockdeep** [comment]
I mentioned this [elsewhere](https://github.com/puma/puma/issues/3028#issuecomment-1350007176), but I noticed that, at least in our tests, the ActiveRecord query cache seems to sometimes be persisting across requests. Maybe that is what is happening here as well?

**Comment by MSP-Greg** [comment]
> at least in our tests

Since you've mentioned testing, we believe that current master should fix the issue.  If you could test against that, it would be appreciated.

**Comment by mockdeep** [comment]
Yeah, it's pretty clearly fixed:

## Puma 6.0.0

![image](https://user-images.githubusercontent.com/80543/207504679-95fd3636-c071-4cad-a8b5-9b266d7c1385.png)

## Puma master

![image](https://user-images.githubusercontent.com/80543/207504644-b66fbf4c-825a-433f-a771-e4d733774175.png)

Any chance of getting a 6.0.1 release soon?

**Comment by MSP-Greg** [comment]
@mockdeep Thank you.  Just wanted clarify...

> Any chance of getting a 6.0.1 release soon?

Yes, PR started, see https://github.com/puma/puma/pull/3032.

**Comment by dentarg** [comment]
Resolved by Puma 6.0.1 I think. Let's us know if that's not the case @JensDebergh

**Comment by JensDebergh** [comment]
@dentarg I've been running it in production for about 5 days now and it looks like the issues have been resolved.

Thank you! 🚀
